### PR TITLE
Migrate to Gradle Kotlin DSL

### PR DESCRIPTION
**Pull request by dbaelz**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [ ] Bug Fix
- [x] Documentation Update

### Minor Documentation Change
Update the link to the "licensing" page in the README.

### Migration to Gradle Kotlin DSL
The Kotlin DSL has some advantages as described in the [documentation](https://docs.gradle.org/current/userguide/kotlin_dsl.html) (e.g. ` enhanced editing experience in supported IDEs, with superior content assist, refactoring [...]`). Everything should work as before, but features like auto-completion in the build scripts are now supported. Furthermore, I added the `buildSrc` directory, which is the [Gradle way](https://docs.gradle.org/current/userguide/organizing_gradle_projects.html#sec:build_sources) to share/abstract imperativ logic, config and so on for build scripts.

The migration is finished, but the build scripts can be improved even more. For example, all version declaration of dependencies could be moved to buildSrc. I didn't changed that because it's optional and I didn't want to make this decision for the project. With https://github.com/thepracticaldev/DEV-Android/pull/40 another PR changes the build scripts. I will integrate these changes when this PR is merged before mine.

Feel free to ask questions or challenge my changes. I hope the migration to the Kotlin DSL is accepted for the project and this PR will be integrated.

## [optional] What gif best describes this PR or how it makes you feel?

Kotlin everywhere

### CODE REVIEW

**Comment by sierisimo** [comment]
This is so awesome! I haven't tried the kotlin DSL in an Android Project but in a kotlin project it's really good. Great PR

**Comment by maestromac** [comment]
Thanks for this @dbaelz !

I took a peak at Kotlin DSL and I'm totally for it. Regarding the buildSrc, what's the benefit of moving all version declaration of dependencies to buildSrc?

 I'll merge this as soon as merge conflict is resolved.

**Comment by sierisimo** [comment]
>  
> I took a peak at Kotlin DSL and I'm totally for it. Regarding the buildSrc, what's the benefit of moving all version declaration of dependencies to buildSrc?
>

I don't see the change of `buildSrc` in this PR but the advantage is that you will have all the versions and string constants in one place, also if the building process require some complex step like copying or editing classes in compilation time, it can be achieved in Kotlin code in classes in `buildSrc` instead of doing patches and hacks on the `build.gradle.kts`

Also the `buildSrc` is compiled **BEFORE** everything so you can add anything that's needed for the building process in it.

But right now… there's no need.

**Comment by sierisimo** [comment]
Heres is an example of it: 
image

There's also de posibility of writing custom tasks for the building process, like pushing changes to somewhere, run custom rules for validation of code, etc. All the steps implied in the building process and quality checks of your code can be achieved programatically in the `buildSrc`

**Comment by dbaelz** [comment]
Thanks @sierisimo you summed it up faster quit well.

The idea (general and in this project) is to avoid redundant information in multiple modules like app, baseui and data. This lowers the risk of bugs due different versions in different modules. It isn't a huge advantage for these simple build scripts, but it's really great if the scripts are more extensive.

We could go even further and for example move parts of the app configuration like "base dependencies" (all dependencies that are redundant defined in all modules) into buildSrc. But for such compact scripts this might be overkill.

I'll updated the PR and integrated the changes of https://github.com/thepracticaldev/DEV-Android/pull/40. Due the renaming of the scripts files I had to readd the changes to the Kotlin build scripts.

File: build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+buildscript {
+    repositories {
+        google()
+        jcenter()
+    }
+    dependencies {
+        classpath("com.android.tools.build:gradle:3.4.1")
+        // Due an issue the full qualified name is required here. See https://github.com/gradle/kotlin-dsl/issues/1291
+        classpath("org.jetbrains.kotlin:kotlin-gradle-plugin:${to.dev.dev_android.build.BuildConfig.kotlinVersion}")
```
Comment (maestromac) [inline build.gradle.kts]: My Gradle sync failed with `Unresolved reference: build` and it's pointing to the fully qualified name.

```
Script compilation errors:

  Line 11:         classpath("org.jetbrains.kotlin:kotlin-gradle-plugin:${to.dev.dev_android.build.BuildConfig.kotlinVersion}")
                                                                          ^ Function invocation 'to(...)' expected

  Line 11:         classpath("org.jetbrains.kotlin:kotlin-gradle-plugin:${to.dev.dev_android.build.BuildConfig.kotlinVersion}")
                                                                          ^ No value passed for parameter 'that'

  Line 11:         classpath("org.jetbrains.kotlin:kotlin-gradle-plugin:${to.dev.dev_android.build.BuildConfig.kotlinVersion}")
                                                                             ^ Unresolved reference: dev

3 errors
```
Did you had to do anything extra to get this to work?

Comment (dbaelz) [reply build.gradle.kts]: No, normally not.  As mention by @sierisimo the buildSrc ist compiled and added to the classpath of every Gradle build script.

You synced it in Android Studio? It worked for me, with Android Studio 3.4 out of the box and without any cleanup. Nevertheless, you could try to execute the clean task once just to be sure that Gradle (and/or your IDE) didn't cache strange stuff and builds the buildSrc. Sometimes AS has a strange behavior when changing larger parts of build scripts (also with Groovy scripts). Reimport of the project, also once just to AS fix itself, might help as well.

Comment (maestromac) [reply build.gradle.kts]: Yeah I synced it, invalidated cache & restart a few times, reimport the project and it still isn't working for me.

Comment (sierisimo) [reply build.gradle.kts]: Android Studio is not 100% to fully support Kotlin DSL (even IntelliJ has it still in an early support). So my steps to check this works is:

1. Check your `gradle/wrapper/gradle-wrapper.properties` is pointing to `distributionUrl=https\://services.gradle.org/distributions/gradle-5.1.1-all.zip`
2. Run the gradle task from the terminal `./gradlew assembleDebug` or something like it. Probably `./gradlew build` will give you the more specific error.

If after checking this two the error still happens is probably a bad idea to go with the Kotlin DSL as it won't work for all developers of the app

Comment (sierisimo) [reply build.gradle.kts]: Worth to say: If your `gradle/wrapper/gradle-wrapper.properties` is not pointing to the right version of gradle, that's a required change in this PR

Comment (dbaelz) [reply build.gradle.kts]: @maestromac Can I help you to solve the problem with the sync? Any update about it?

Comment (maestromac) [reply build.gradle.kts]: Sorry for the late reply here. @sierisimo your step one points to gradle-5.1.1 while this PR updated it to 5.4.1. Mine is already pointed to 5.4.1 and it doesn't work. I also can't get it to work on my windows machine, same error. Output [here](https://gist.github.com/maestromac/9165abd839ff2f85700bdbc0d7776d1e)

Comment (dbaelz) [reply build.gradle.kts]: Oh, I've failed hard. Due the gitignore of the project the buildSrc/src folder was ignored. So I didn't commit the Config.kt and unfortunately oversee it after your comments. I changed the gitignore and commited the file.

Sorry for the inconvenience and a giphy for all the trouble.

Failed

Comment (maestromac) [reply build.gradle.kts]: That's alright 😄 . It's working for me now.

**Pull request merged**